vowpalwabbit/cache.h

```diff
@@ -16,3 +16,18 @@ void cache_tag(io_buf& cache, v_array<char> tag);
 void cache_features(io_buf& cache, example* ae, uint64_t mask);
 void output_byte(io_buf& cache, unsigned char s);
 void output_features(io_buf& cache, unsigned char index, features& fs, uint64_t mask);
+
+namespace VW
+{
+  template <typename T>
+  T read_object(io_buf& cache)
+  {
+    char* c;
+    size_t next_read_size = sizeof(T);
+    if (cache.buf_read(c, next_read_size) < next_read_size)
+      THROW("Failed to read CCB label cache");
```

Comment:
Maybe we should have a return reference?  The template will work better and we can return the number of bytes read (effectively, an error code) rather than throwing.

Reply:
The pollutes usage of this function a lot. Can we actually recover by knowing number of bytes read if we fail reading at this point?

Reply:
The cache code is generally setup so that failure to read => end of cache => end pass.  Letting an error create this semantics seems reasonable?

Reply:
Sounds good, I'll fix that up

Reply:
Essentially all of the helpfulness of this function is lost when the return needs to be an out param (see below). To be fair though the semantics do change a bit. Since the code path isn't setup to handle exceptions to indicate that the cache file has ended or there is a failure then it seems like the function I proposed is not really an option.

```c++
auto size_probs = read_object<uint32_t>(cache);
read_count += sizeof(size_probs);

```
---
```c++
uint32_t size_probs;
if(read_object(cache, size_probs) < sizeof(size_probs))
  return 0;
read_count += sizeof(size_probs);
```

Reply:
I see.   I'm not sure what's better than the last.  Maybe a macro?  Or maybe just give up on read_object.

Reply:
Yeah I think I'll give up on read_object and just stick with the current convention. It didn't reduce code that much anyway